yield_plugin/src/yield_plugin.cpp

```diff
@@ -403,106 +415,243 @@ namespace yield_plugin
+
+    // Iterate through the object to check if it's on the route
+    bool on_route = false;
+    int on_route_idx = 0;
+
+    for (int j = 0; j < trajectory2.size(); ++j) 
+    {
+      lanelet::BasicPoint2d curr_point;
+      curr_point.x() = trajectory2[j].predicted_position.position.x;
+      curr_point.y() = trajectory2[j].predicted_position.position.y;
+
+      auto corresponding_lanelets = wm_->getLaneletsFromPoint(curr_point, 4); // get 4 lanelets
+
+      for (auto llt: corresponding_lanelets)
+      {
+        RCLCPP_DEBUG_STREAM(rclcpp::get_logger("yield_plugin"), "Checking llt: " << llt.id());
+
+        if (route_llt_ids_.find(llt.id()) != route_llt_ids_.end())
+        {
+          on_route = true;
+          on_route_idx = j;
+          break;
+        }
+      }
+      if (on_route)
+        break;
+    }
+
+    if (!on_route)
+    {
+      RCLCPP_DEBUG_STREAM(rclcpp::get_logger("yield_plugin"), "Predicted states are not on the route! ignoring");
+      return boost::none;
+    }
+    
+    double smallest_dist = std::numeric_limits<double>::max();
+    for (int i = 0; i < trajectory1.trajectory_points.size() - 1; ++i) 
+    {
+      auto p1a = trajectory1.trajectory_points[i];
+      auto p1b = trajectory1.trajectory_points[i + 1];
+      for (int j = on_route_idx; j < trajectory2.size() - 1; ++j) 
+      {
+        auto p2a = trajectory2[j];
+        auto p2b = trajectory2[j + 1];
+        double p1a_t = rclcpp::Time(p1a.target_time).seconds();
+        double p1b_t = rclcpp::Time(p1b.target_time).seconds();
+        double p2a_t = rclcpp::Time(p2a.header.stamp).seconds();
+        double p2b_t = rclcpp::Time(p2b.header.stamp).seconds();
+
+        RCLCPP_DEBUG_STREAM(rclcpp::get_logger("yield_plugin"), "p1a.target_time: " << std::to_string(p1a_t) << ", p1b.target_time: " << std::to_string(p1b_t));
+        RCLCPP_DEBUG_STREAM(rclcpp::get_logger("yield_plugin"), "p2a.target_time: " << std::to_string(p2a_t) << ", p2b.target_time: " << std::to_string(p2b_t));
+        RCLCPP_DEBUG_STREAM(rclcpp::get_logger("yield_plugin"), "p1a.x: " << p1a.x << ", p1a.y: " << p1a.y);
+        RCLCPP_DEBUG_STREAM(rclcpp::get_logger("yield_plugin"), "p1b.x: " << p1b.x << ", p1b.y: " << p1b.y);
+
+        RCLCPP_DEBUG_STREAM(rclcpp::get_logger("yield_plugin"), "p2a.x: " << p2a.predicted_position.position.x << ", p2a.y: " << p2a.predicted_position.position.y);
+        RCLCPP_DEBUG_STREAM(rclcpp::get_logger("yield_plugin"), "p2b.x: " << p2b.predicted_position.position.x << ", p2b.y: " << p2b.predicted_position.position.y);
+
+        // Linearly interpolate positions at a common timestamp for both trajectories
+        double dt = (p2a_t - p1a_t) / (p1b_t - p1a_t);
+        double x1 = p1a.x + dt * (p1b.x - p1a.x);
+        double y1 = p1a.y + dt * (p1b.y - p1a.y);
+        double x2 = p2a.predicted_position.position.x;
+        double y2 = p2a.predicted_position.position.y;
```

Comment:
This should ideally be moved to a separate function. We can keep this as is for now, but I wanted to make a note.